### PR TITLE
[core] Incremental offline downloads

### LINKED ISSUE #4414: Downloading an offline pack blocks the visible map from loading
If the active style of a map view is changed while an offline pack is downloading, the new style will fail to load (i.e., it only shows a black screen) while that download is active.

1. Start downloading an offline pack
2. Switch the visible style on your map view
3. The new style will not load until the offline pack finishes (or pauses) downloading

/cc @jfirebaugh @1ec5 @zugaldia

### PR DESCRIPTION
**Pull request by jfirebaugh**
Don't allow OfflineDownload to flood the request queue. Fixes #4414.

### CODE REVIEW

**Comment by tobrun** [comment]
I was thinking of picking up this `gl-school` ticket. Really cool to see it happening and need to admit I wouldn't have come up with same code. 👀  and 📓 ✏️ 'ing.

File: platform/default/mbgl/storage/offline_download.cpp [hunk 1]
```diff
@@ -217,30 +182,56 @@ void OfflineDownload::activateDownload() {
...
         }
 
-        for (const auto& resource : glyphResources(parser)) {
-            ensureResource(resource);
+        if (!parser.spriteURL.empty()) {
+            queueResource(Resource::spriteImage(parser.spriteURL, definition.pixelRatio));
+            queueResource(Resource::spriteJSON(parser.spriteURL, definition.pixelRatio));
         }
+
+        continueDownload();
```
Comment (tmpsantos) [inline platform/default/mbgl/storage/offline_download.cpp]: Shouldn't we also call `continueDownload()` on error?

File: platform/default/mbgl/storage/offline_download.cpp [hunk 2]
```diff
@@ -299,9 +286,7 @@ void OfflineDownload::ensureResource(const Resource& resource,
                 return;
             }
 
-            if (status.complete()) {
-                setState(OfflineRegionDownloadState::Inactive);
-            }
+            continueDownload();
```
Comment (tmpsantos) [inline platform/default/mbgl/storage/offline_download.cpp]: Shouldn't we call `continueDownload()` also on error?

Comment (jfirebaugh) [inline platform/default/mbgl/storage/offline_download.cpp]: A request that errors will be retried after some delay. So in that sense it's still "active" and consuming resources, notably the request object, its timer, and network resources when the timer fires.

We _could_ try to squeeze in subsequent requests while we wait for the errored request to retry. But that risks overloading the upstream request queue -- defeating our own metering here -- if there are a lot of errored requests that all come up for retry at the same time.

Many times, the cause of a request error will apply to many requests of the same type. For instance if a server is unreachable, all the requests to that host are going to error. In that case, continuing to try subsequent resources after the first few errors is fruitless anyway.

I'll add the above rationale as a comment.

File: platform/default/mbgl/storage/offline_download.cpp [hunk 3]
```diff
@@ -152,18 +113,21 @@ OfflineRegionStatus OfflineDownload::getStatus() const {
         }
     }
 
-    result.requiredResourceCount += spriteResources(parser).size();
-    result.requiredResourceCount += glyphResources(parser).size();
+    if (!parser.glyphURL.empty()) {
+        result.requiredResourceCount += parser.fontStacks().size() * 256;
```
Comment (1ec5) [inline platform/default/mbgl/storage/offline_download.cpp]: Factor 256 (used multiple times here) into a global constant?

**Pull request merged**